### PR TITLE
feat: Script mode support for Estimator class

### PR DESCRIPTION
**Pull request by evakravi**
*Issue #, if available:*

*Description of changes:*
This PR adds script mode support to the `Estimator` and `EstimatorBase` classes. 

This was basically done by adding the parameters `source_dir, git_config, hyperparameters, container_log_level, code_location, entry_point, dependencies` to the `Estimator` and `EstimatorBase` classes. 

*Testing done:*
The changes to the `Estimator` class do not break any existing unit tests. In addition, new unit tests were added to simulate the script mode use case for the `Estimator` class, and confirm that the calls to `sagemaker.create_training_job()` and `s3` are the same for the `Estimator` class and `Framework` class when both use script mode. A test was also added to ensure that git support works with the `Estimator` class. Integration tests will be introduced in a subsequent PR. 

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I certify that the changes I am introducing will be backword compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [x] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [x] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by JGuinegagne — CHANGES_REQUESTED** [review]
Would you be able to add unit tests to verify that:
- `git_config` is also handled correctly and similarly b7 `DummyFramework` and `Estimator`
- `hyperparameters` are handled correctly (if not covered already)

File: src/sagemaker/estimator.py [hunk 1]
```diff
@@ -2376,6 +2690,10 @@ def _model_entry_point(self):
 
         return None
 
+    def set_hyperparameters(self, **kwargs):
+        """Sets hyperparameters."""
```
Comment (JGuinegagne) [inline src/sagemaker/estimator.py:2694]: Let's avoid repeating function name in docstring. How about:
```py
"""Escape the dict argument as JSON, update the private hyperparameter attribute."""
```

File: src/sagemaker/model.py [hunk 2]
```diff
@@ -399,12 +399,12 @@ def prepare_container_def(
...
-            elif self.source_dir and self.entry_point:
-                self._upload_code(deploy_key_prefix, repack=True)
-            else:
-                self._upload_code(deploy_key_prefix, repack=False)
+            self._upload_code(
+                deploy_key_prefix,
+                repack=self.source_dir
+                and self.entry_point
+                and not (self.key_prefix or self.git_config),
+            )
```
Comment (JGuinegagne) [inline src/sagemaker/model.py:407]: nit for readability, i would vote for:
```py
is_repack = self.source_dir and self.entry_point and not (self.key_prefix or self.git_config)
self._upload_code(deploy_key_prefix, repack=is_repack)
```

**Review by JGuinegagne — COMMENTED** [review]
A few non-blocking comments/suggestions.

File: src/sagemaker/estimator.py [hunk 3]
```diff
@@ -456,10 +616,105 @@ def _prepare_for_training(self, job_name=None):
...
+        elif local_mode:
+            code_bucket, key_prefix = parse_s3_url(self.code_location)
+            code_s3_prefix = "/".join(filter(None, [key_prefix, self._current_job_name, "source"]))
+            kms_key = None
+        else:
+            code_bucket, key_prefix = parse_s3_url(self.code_location)
+            code_s3_prefix = "/".join(filter(None, [key_prefix, self._current_job_name, "source"]))
+
+            output_bucket, _ = parse_s3_url(self.output_path)
+            kms_key = self.output_kms_key if code_bucket == output_bucket else None
```
[the remark this answers is not in the document]
Comment (evakravi) [reply src/sagemaker/estimator.py:705]: No idea, this section was directly lifted from somewhere else in the code.

**Review by JGuinegagne — APPROVED** [review]

File: src/sagemaker/estimator.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (JGuinegagne) [reply src/sagemaker/estimator.py:705]: @shreyapandit @mufaddal-rohawala ?

**Review by mufaddal-rohawala — APPROVED** [review]

File: src/sagemaker/estimator.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (shreyapandit) [reply src/sagemaker/estimator.py:705]: will need to dive into this.

**Review by shreyapandit — APPROVED** [review]

**Pull request merged**